### PR TITLE
[Intel MKL] Changes in common files to enable MKL Quantized ops with native format

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**
This PR adds the common changes to enable MKL Quantized ops in native format mode. It removes the dependency on the duplicate number of inputs/outputs for quantized ops.

Added by @penpornk for API review:
* The ops that are changed in API Golden files all have hidden visibility.
* The attribute `data_format` is added because MKL ops support both TF's native NHWC format and its own blocked format.
* The attributes `is_weight_const` and `is_filter_const` are added for caching purposes.
* The changes should be backward compatible.

### CODE REVIEW

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm so sorry for the delay! The overall PR looks good to me. 
I'm a bit hesitant on adding oneDNN-specific parameters (`is_*_const`) to the non-MKL-prefixed Quantized ops (which are visible to C++ API users). But I'll leave this part to the API review. (Will tag them after I approve this PR.)

File: tensorflow/core/common_runtime/mkl_layout_pass.cc [hunk 1]
```diff
@@ -2849,7 +2855,9 @@ void MklLayoutRewritePass::CopyAttrsQuantizedConv2D(const Node* orig_node,
   nb->Attr("is_filter_const", filter_node->IsConstant());
   nb->Attr("strides", strides);
   nb->Attr("dilations", dilations);
-  nb->Attr("T", out_type);  // added "T" for facilitating MklToTf conversion.
+  if (!NativeFormatEnabled()) {
+    nb->Attr("T", out_type);  // added "T" for facilitating MklToTf conversion.
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/common_runtime/mkl_layout_pass.cc]: Done.

File: tensorflow/core/graph/mkl_graph_util.h [hunk 2]
```diff
@@ -144,9 +145,27 @@ static const char* const kMklEagerOpPrefix = "_MklEager";
...
+}
+
 // Get the name of Mkl Native (does not depend on layout propagation) op
 // from original TensorFlow op.
 inline string GetMklNativeOpName(const string& name) {
+  // For quantized ops in native format mode, we won't change the op name and
+  // will just reuse original name of the op if we own that name (the one
+  // witout _Mkl prefix). However, we may change the attributes for such ops.
+  // There are few quantized ops where we don't own original name. Those will
+  // be prfixed with _MklNative just like native ops.
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/graph/mkl_graph_util.h]: Done.

File: tensorflow/core/framework/common_shape_fns.cc [hunk 3]
```diff
@@ -2452,6 +2452,59 @@ Status SparseReduceShapeFn(InferenceContext* c) {
...
+    DimensionHandle depth;
+    TF_RETURN_IF_ERROR(
+        c->Merge(c->Dim(minmax, 0), c->Dim(input, axis), &depth));
+  }
+  c->set_output(1, minmax);
+  c->set_output(2, minmax);
+  return Status::OK();
+}
+
+Status DequantizeShape(InferenceContext* c) {
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/framework/common_shape_fns.cc]: Done.

**Comment by mahmoud-abuzaina** [comment]
Thank you for reviewing the PR. I have made the requested changed.
For adding an attribute to the ops, I thought these ops are not visible in the API as we mark them as hidden, for e.g. [here](https://github.com/tensorflow/tensorflow/blob/5892c5bbe5edc62b1f30c6cf005a5f2e37a293b4/tensorflow/core/api_def/base_api/api_def_QuantizedDepthwiseConv2DWithBias.pbtxt#L3). These ops usually get used by frozen graphs generated by our offline tool.

**Review by penpornk — APPROVED** [review]
Thank you for the fixes!

> For adding an attribute to the ops, I thought these ops are not visible in the API as we mark them as hidden, for e.g. [here](https://github.com/tensorflow/tensorflow/blob/5892c5bbe5edc62b1f30c6cf005a5f2e37a293b4/tensorflow/core/api_def/base_api/api_def_QuantizedDepthwiseConv2DWithBias.pbtxt#L3).

Oh, you're right. Sorry I forgot to check before. :)

However, this PR will still require an API review anyway because it modifies the API golden files (`//tensorflow/tools/api/golden/...`). Tagging API review now. (It will be reviewed next week.)

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 4]
```diff
@@ -1994,6 +1976,54 @@ Uses oneDNN APIs to perform fused batch normalization and relu.
...
+    .Input("min_input: float")
+    .Input("max_input: float")
+    .Input("min_filter: float")
+    .Input("max_filter: float")
+    .Output("output: out_type")
+    .Output("min_output: float")
+    .Output("max_output: float")
+    .Attr("Tinput: quantizedtype")
+    .Attr("Tfilter: quantizedtype")
+    .Attr("T: quantizedtype")
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:1991]: Is `T` needed? Same for other native ops in this PR.

**Comment by gbaned** [comment]
@mahmoud-abuzaina Can you please check @penpornk's comments and keep us posted ? Thanks!

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 4 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: It is not needed for this particular op as this `T` is needed for Conv only in non-native path to enable MklToTF conversion. Required datatypes for Conv are named differently (Tfilter, Tbias). So removed it. However, in other ops like Max/AvgPooling, `T` is the required datatype for input/output. Thank you for pointing out this!

Comment (penpornk) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Thank you for the explanation!

**Review by penpornk — APPROVED** [review]
Thank you for the change!

**Review by penpornk — CHANGES_REQUESTED** [review]
@mahmoud-abuzaina Sorry for the delay! The API review committee would like to see how the added attributes will be used. Can we put these attribute changes in the PRs that actually have the implementation that use them? We can keep other general changes in this PR.

**Comment by mahmoud-abuzaina** [comment]
@penpornk no worries. This PR and other relative pending PRs don't add any new implementation to use the newly added attributes. Rather, the implementation that uses these attributes has been there for long time. 
Here is the background: all Quantized* ops are hidden API ops and get generated/added to frozen pb file via offline tool. They were considered dummy ops. At runtime those ops will be rewritten to corresponding _MklQuantized*. The purpose of _MklQuantized* is to handle the extra number of inputs and outputs that we used to require. With native format support, we are getting rid of all extra inputs and outputs. So I thought we can reuse Quantized* ops (that are usable only via oneDNN integration code) instead of creating new ops. While doing that I noticed that these ops are missing few required attributes. These attributes have been used for a while with existing implementation, but the missing attributes were not causing a failure before because Quantized* ops were dummy ops and will be always rewritten to _Mkl ones which have the those attributes. Now since we are using Quantized* named ops, we are adding the missing attributes but they will use existing implementation. As an example of existing attribute usage see [here](https://github.com/tensorflow/tensorflow/blob/697f117f36176fc5d07c0b37d1b17b9d7e3a3bc9/tensorflow/core/kernels/mkl/mkl_conv_ops.cc#L697). Sorry for my verbosity and please let me if I am missing something.

**Comment by mahmoud-abuzaina** [comment]
Closing for now. I will do some changes and reopen once ready.

**Comment by penpornk** [comment]
@mahmoud-abuzaina So sorry for my delayed reply and thank you very much for the thorough explanation! I didn't remember that these attributes are already used in the already checked-in kernels. 

Re: Hidden ops: The API review committee reminded me that op names not prefixed by `_` are still callable. Marking them hidden just puts them in [tf.raw_ops](https://www.tensorflow.org/api_docs/python/tf/raw_ops), so they are still part of the API golden.

**Review by penpornk — APPROVED** [review]
I'm so sorry for the long delay. Thank you again for the PR!

**Pull request merged**